src/sage/geometry/cone_critical_angles.py

```diff
@@ -0,0 +1,998 @@
+        sage: GG = G.transpose()*G
+        sage: hs = [h.change_ring(AA).normalized() for h in Q]
+        sage: H = matrix.column(hs)
+        sage: HH = H.transpose()*H
+        sage: M = G.transpose()*H
+        sage: G_index_sets = list(gevp_licis(G))
+        sage: H_index_sets = list(gevp_licis(H))
+        sage: all(
+        ....:   set(
+        ....:     _normalize_gevp_solution(s)
+        ....:     for s in
+        ....:     chain(
+        ....:       solve_gevp_zero(M,I,J),
+        ....:       solve_gevp_nonzero(GG,HH,M,I,J)
+        ....:     )
+        ....:   )
+        ....:   ==
+        ....:   set(
+        ....:     _normalize_gevp_solution(s)
+        ....:     for s in
+        ....:     _solve_gevp_naive(GG,HH,M,I,J)
+        ....:   )
+        ....:   for I in G_index_sets
+        ....:   for J in H_index_sets
+        ....: )
+        True
+
+    According to Proposition 7, the only eigenvalues that arise when
+    either ``G`` or ``H`` is invertible are `-1`, `0`, and `1`::
+
+        sage: # long time
+        sage: from sage.geometry.cone_critical_angles import (
+        ....:   _random_admissible_cone,
+        ....:   gevp_licis,
+        ....:   solve_gevp_nonzero)
+        sage: n = ZZ.random_element(1,3)
+        sage: P = _random_admissible_cone(ambient_dim=n)
+        sage: Q = _random_admissible_cone(ambient_dim=n)
+        sage: gs = [g.change_ring(AA).normalized() for g in P]
+        sage: hs = [h.change_ring(AA).normalized() for h in Q]
+        sage: G = matrix.column(gs)
+        sage: GG = G.transpose()*G
+        sage: H = matrix.column(hs)
+        sage: HH = H.transpose()*H
+        sage: M = G.transpose()*H
+        sage: from itertools import product
+        sage: all(
+        ....:  (v in [-1,0,1]
+        ....:   for (v,_,_,_) in solve_gevp_nonzero(GG,HH,M,I,J))
+        ....:   for (I,J) in product(gevp_licis(G),gevp_licis(H))
+        ....:   if len(I) == n or len(J) == n )
+        True
+    """
+    if len(J) < len(I):
+        # We can always opt to solve the smaller problem. Reading the
+        # first three assignments below, you should be able to
+        # convince yourself that switching GG <-> HH, I <-> J, and
+        # transposing M does in fact switch from the "xi problem" to
+        # the "eta problem."
+        yield from ( (l, xi, eta, m)
+                     for (l, eta, xi, m)
+                     in solve_gevp_nonzero(HH,GG,M.transpose(),J,I) )
+    else:
+        G_I_pinv_H_J = GG[I,I].inverse_positive_definite()*M[I,J]
+        H_J_pinv_G_I = HH[J,J].inverse_positive_definite()*M.transpose()[J,I]
+        L = (G_I_pinv_H_J * H_J_pinv_G_I)
+
+        for (sigma, xis, m) in L.eigenvectors_right():
+            for xi in xis:
+                if sigma > 0:
+                    for l in [-sigma.sqrt(), sigma.sqrt()]:
+                        eta = ~l * H_J_pinv_G_I*xi
+                        eta.set_immutable()
+                        yield (l, xi, eta, m)
```

Comment:
```suggestion
        for (sigma, xis, m) in L.eigenvectors_right():
            if sigma > 0:
                sigma_sqrt = sigma.sqrt()
                inv_sqrt = ~sigma_sqrt
                sqrts = [-sigma_sqrt, sigma_sqrt]
                inv_sqrts = [-inv_sqrt, inv_sqrt]
                for xi in xis:
                    for l, li in zip(sqrts, inv_sqrts):
                        eta = li * H_J_pinv_G_I * xi
                        eta.set_immutable()
                        yield (l, xi, eta, m)
```
Trying to avoid repeating computations.

Reply:
Do you think it's worth pulling out the `zip` as well? Something like,

```python
pm_sqrt_inv_pairs = [
    (-sigma_sqrt, -inv_sqrt),
    (sigma_sqrt, inv_sqrt),
]
for xi in xis:
    for l, li in pm_sqrt_inv_pairs:
```

I can't say for sure now that `zip` returns an iterator.

Reply:
Avoiding `zip` looks like it saves about a microsecond, but at that scale it's impossible to say for sure.

Reply:
Done, thanks

Reply:
Pulling out the zip wouldn't work because of the iterator, but you're right that it should be better to just do it manually since it is a very small list. In either case, the biggest thing is making sure the `if sigma > 0:` is outside the `for` loops.